control-plane/subcommand/acl-init/command.go

```diff
@@ -162,47 +137,31 @@ func (c *Command) Run(args []string) int {
 	}
 
 	var secret string
-	if c.flagACLAuthMethod != "" {
-		cfg := api.DefaultConfig()
-		c.http.MergeOntoConfig(cfg)
+	if c.consul.ConsulLogin.AuthMethod != "" {
+		serverConnMgrConfig, err := c.consul.ConsulServerConnMgrConfig()
+		if err != nil {
+			c.logger.Error("Failed to construct consul server connection manager", "error", err)
+		}
 
-		if len(c.flagServerAddresses) > 0 {
-			serverAddresses, err := common.GetResolvedServerAddresses(c.flagServerAddresses, c.providers, c.logger)
+		if c.watcher == nil {
+			c.watcher, err = discovery.NewWatcher(c.ctx, serverConnMgrConfig, c.logger.Named("consul-server-connection-manager"))
 			if err != nil {
-				c.UI.Error(fmt.Sprintf("Unable to discover any Consul addresses from %q: %s", c.flagServerAddresses[0], err))
-				return 1
-			}
-
-			scheme := "http"
-			if c.flagUseHTTPS {
-				scheme = "https"
+				c.logger.Error("Failed to construct Consul server watcher", "error", err)
 			}
-
-			serverAddr := fmt.Sprintf("%s:%d", serverAddresses[0], c.flagServerPort)
-			cfg.Address = serverAddr
-			cfg.Scheme = scheme
 		}
 
-		c.consulClient, err = consul.NewClient(cfg, c.http.ConsulAPITimeout())
-		if err != nil {
-			c.logger.Error("Unable to get client connection", "error", err)
-			return 1
-		}
+		go c.watcher.Run()
+		defer c.watcher.Stop()
 
-		loginParams := common.LoginParams{
-			AuthMethod:      c.flagACLAuthMethod,
-			Datacenter:      c.flagPrimaryDatacenter,
-			BearerTokenFile: c.bearerTokenFile,
-			TokenSinkFile:   c.flagTokenSinkFile,
-			Meta: map[string]string{
-				"component": c.flagComponentName,
-			},
+		c.consulClient, err = consul.NewClientFromConnMgr(c.consul.ConsulClientConfig(), c.watcher)
+		if err != nil {
+			c.logger.Error("Failed to create Consul client", "error", err)
 		}
-		secret, err = common.ConsulLogin(c.consulClient, loginParams, c.logger)
+		state, err := c.watcher.State()
 		if err != nil {
-			c.logger.Error("Consul login failed", "error", err)
-			return 1
+			c.logger.Error("Failed to get Consul watcher state", "error", err)
 		}
+		secret = state.Token
```

Comment:
I'd love a suggestion here with regards to finding a way to get the token without having to read the `State` again. I could use the variant of `NewClientFromConnManager` that accepts that watcher state but I think we want to move away from using that method directly.

Reply:
I believe that was pretty much how Iryna had to do it too in the sync-catalog PR @ https://github.com/hashicorp/consul-k8s/blob/main/control-plane/subcommand/sync-catalog/command.go#L232